Ruby Capstone project

Kccendana opened the pull request:
**### Project requirements**


- Each class from the UML class diagram should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- All methods visible in the diagram should be implemented:
   - add_item method (in all classes that have that method)
     - should take an instance of the Item class as an input.
     - should add the input item to the collection of items.
     - should add self as a property of the item object (by using the correct setter from the item object).
  - can_be_archived?() in the Item class
     - should return true if published_date is older than 10 years.
     - otherwise, it should return false.
  - move_to_archive() in the Item class
    - should reuse can_be_archived?() method.
    - should change the archived property to true if the result of the can_be_archived?() method is true.
    - should do nothing if the result of the can_be_archived?() method is false.
  - can_be_archived?() in the Book class
     - should override the method from the parent class.
     - should return true if parent's method returns true OR if cover_state equals to "bad".
     - otherwise, it should return false.
  - can_be_archived?() in the MusicAlbum class
     - should override the method from the parent class.
     - should return true if parent's method returns true AND if on_spotify equals true.
     - otherwise, it should return false.
  - can_be_archived?() in the Movie class
     - should override the method from the parent class.
     - should return true if parent's method returns true OR if silent equals true.
     - otherwise, it should return false.
  - can_be_archived?() in the Game class
     - should override the method from the parent class.
     - should return true if parent's method returns true AND if last_played_at is older than 2 years.
     - otherwise, it should return false.
- Add unit tests for all implemented methods.
### User interface
- Create a main.rb file that will serve as your console app entry-point.
- Your console app, at the start, should:
  - Present the user with a list of options to perform.
  - Let users choose an option.
  - If needed, ask for parameters for the option.
  - Have a way to quit the app.
- The following options should be available:
  - List all books
  - List all music albums
  - List all movies
  - List of games
  - List all genres (e.g 'Comedy', 'Thriller')
  - List all labels (e.g. 'Gift', 'New')
  - List all authors (e.g. 'Stephen King')
  - List all sources (e.g. 'From a friend', 'Online shop')
  - Add a book
  - Add a music album
  - Add a movie
  - Add a game
- All data should be preserved by saving collections in .json files.
###  Database
- Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### Here is the link for our video presentation

[Video Demo presentation](https://drive.google.com/file/d/1UK-VH0knX5Ao--2zQwLXuqCIie0daD1N/view?usp=sharing)

Kccendana committed:
add initial files and linter configuration

Commit:
Ft project-setup

Commit:
Add main.rb as entry point
Add menu.rb as SOLID implementation and DRY codes
Add catalog_management to handle all functionality of application

# Finishes ..

Commit:
team-2

Add genres class
Add test of genres class,
Add music_album classs
Add test of genres class
Add method in catalog do add music-album and genres
Add method definitions to insert data into those array

Commit:
Add data to json file and testing

Kccendana committed:
Create add_item_info.rb

Kccendana committed:
Create book.rb

Kccendana committed:
Create books.json

Kccendana committed:
Update catalog_management.rb

Kccendana committed:
Update genres.rb

Kccendana committed:
Update inc_helper.rb

Kccendana committed:
Update item_class.rb

Kccendana committed:
Update main.rb

Kccendana committed:
Create label.rb

Kccendana committed:
Update menu.rb

Kccendana committed:
Update music_album.rb

Kccendana committed:
Create music_albums.json

Kccendana committed:
Update genre_rspec.rb

Kccendana committed:
Update item_rspec.rb

Commit:
Fix rubocop errors

Kccendana committed:
Merge pull request #14 from Kccendana/team-1

Team 1

Commit:
Fix loading for the book.json

Kccendana committed:
modify

Kccendana committed:
modify books and label

Commit:
Fixing the bugs and issue of the first branch

Commit:
Fix linters error

zbahati committed:
Merge pull request #16 from Kccendana/musicAlbum

Fixing the bugs and issue of the first branch

Commit:
Add Genre name

Commit:
Add schema for music and genre

zbahati committed:
Merge pull request #17 from Kccendana/schema_music_genre

Add schema for music and genre

Nelofarzabi committed:
Add feature team 3b

Nelofarzabi committed:
Remove some files

Nelofarzabi committed:
Fix linters

Kccendana committed:
Update book.rb

Kccendana committed:
Update add_item_info.rb

Kccendana committed:
Update books.json

Kccendana committed:
Update catalog_management.rb

Kccendana committed:
Update inc_helper.rb

Kccendana committed:
Update item_class.rb

Kccendana committed:
Update label.rb

Kccendana committed:
Create labels.json

Kccendana committed:
added labels and book to the menu

Kccendana committed:
add tests for book label

Kccendana committed:
Merge branch 'develop' into book-label

Kccendana committed:
fix linter error

Kccendana committed:
fix linter error

Kccendana committed:
modify database

Commit:
Fix linters error and testing part

zbahati committed:
Merge pull request #18 from Kccendana/team-3b

Create Game and Author Class

zbahati committed:
Merge branch 'develop' into book-label

Commit:
Fix rubocop errors

Commit:
Merge branch 'book-label' of https://github.com/Kccendana/catalog_of_my_things into book-labe

Commit:
ft book-label

Fixing merging conflict
Fixing linters errors
Adding the Module to handle rubocop erros
Fix Bugs when combining all of the files when submiting the book-label

zbahati committed:
Merge pull request #15 from Kccendana/book-label

Book label

Commit:
Fix the License link in Readme

Commit:
Fix the Options in the menu and set the quit to be 12 in main

Kccendana committed:
add video presentation

frank1738 reviewed (changes requested):
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media4.giphy.com/media/r5eafWP6dcbxX8cLeN/giphy.gif"/>


## Highlights

✅ Your presentation is top-notch! 👍 
✅ Demonstrated effective error handling in challenging scenarios.
✅ Successfully fulfilled the majority of the project requirements. Well done! 👍🌟👍 
✅ Hats off to you for crafting clean, readable, and easily maintainable code! 🙌👍


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Commit:
Ft develop

Add publisher to the book class
Test book with all function
Test genre with add_item method
Test label with add_item method

# Finishes ..

Commit:
Merge branch 'develop' of https://github.com/Kccendana/catalog_of_my_things into develop

Whoistolu commented on book.rb:
```diff
@@ -0,0 +1,31 @@
+    @cover_state = cover_state
+    @archived = can_be_archived?
+    @label = label
+  end
```
- Well resolved

Whoistolu commented on test/book_rspec.rb:
```diff
@@ -0,0 +1,24 @@
+      expect(new_book.cover_state).to eq('bad')
+      expect(new_book.archived).to eq(true)
+    end
+  end
```
- well resolved

Whoistolu commented on test/genre_rspec.rb:
```diff
@@ -0,0 +1,19 @@
+  it 'Should be associated with an item' do
+    expect(@item.genre).to eq(@genre)
+  end
+end
```
- well resolved

Whoistolu commented on test/label_rspec.rb:
```diff
@@ -0,0 +1,23 @@
+    expect(new_label.title).to eq('JSON Label')
+    expect(new_label.color).to eq('Green')
+  end
+end
```
- well resolved

Whoistolu reviewed (approved):
### Hi Team,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

# STATUS: APPROVED 🟢
**Congratulations! 🎉** 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
